File: frontend/src/modules/specialty/types/childrens.ts
/**
 * TypeScript types for the Children's Book Studio.
 *
 * Covers book creation wizard, page spread editor, character consistency,
 * text readability, bilingual support, safety & provenance, and export.
 *
 * Blueprint refs: 3.1-3.9, 13.1, 14.1
 */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export type AgeRange = "board" | "picture" | "early_reader" | "chapter";

export type IllustrationStyle =
  | "watercolor"
  | "cartoon"
  | "flat"
  | "storybook"
  | "realistic"
  | "crayon_pencil"
  | "collage"
  | "anime_manga";

export type ColorPalette =
  | "bright_vibrant"
  | "soft_pastel"
  | "warm_earthy"
  | "cool_dreamy"
  | "monochrome_accent";

export type StoryMode = "prose" | "rhyming" | "repetitive_cumulative";

export type CreationMode = "ai_generate" | "write_own" | "import_text";

export type BilingualLayout = "side_by_side" | "alternating" | "back_section";

export type FearIntensity = "none" | "mild" | "moderate";

export type Tone = "warm_reassuring" | "exciting" | "humorous" | "educational";

export type PageLayout =
  | "full_bleed"
  | "top_image_bottom_text"
  | "bottom_image_top_text"
  | "left_image_right_text"
  | "right_image_left_text"
  | "text_only"
  | "full_bleed_no_text";

export type TextPosition = "top" | "middle" | "bottom";

export type BookStatus = "draft" | "in_progress" | "published";

export type PageType = "story" | "title" | "dedication" | "credits" | "blank";

// ---------------------------------------------------------------------------
// Core interfaces
// ---------------------------------------------------------------------------

export interface ChildrensBook {
  id: string;
  org_id: string;
  title: string;
  subtitle?: string;
  author: string;
  age_range: AgeRange;
  page_count: number;
  trim_size: string;
  illustration_style: IllustrationStyle;
  color_palette: ColorPalette;
  story_mode: StoryMode;
  creation_mode: CreationMode;
  is_bilingual: boolean;
  bilingual_language?: string;
  bilingual_layout?: BilingualLayout;
  fear_intensity: FearIntensity;
  tone: Tone;
  theme_moral?: string;
  main_character?: string;
  setting?: string;
  story_prompt?: string;
  status: BookStatus;
  qa_score?: number;
  cover_image_url?: string;
  created_at: string;
  updated_at: string;
}

export interface ChildrensBookPage {
  id: string;
  book_id: string;
  page_number: number;
  page_type: PageType;
  layout: PageLayout;
  text_content?: string;
  translated_text?: string;
  text_font?: string;
  text_size?: number;
  text_color?: string;
  text_position?: TextPosition;
  text_plate_enabled: boolean;
  illustration_prompt?: string;
  illustration_url?: string;
  illustration_model?: string;
  illustration_seed?: string;
  contrast_score?: number;
  gutter_safe: boolean;
  created_at: string;
  updated_at: string;
}

export interface ChildrensBookCharacter {
  id: string;
  book_id: string;
  name: string;
  species?: string;
  description: string;
  reference_images: string[];
  auto_append: boolean;
  clothing_rules?: Record<string, unknown>;
  scale_rules?: Record<string, unknown>;
  setting_continuity_rules?: Record<string, unknown>;
  time_of_day_rules?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

// ---------------------------------------------------------------------------
// Wizard / creation payloads
// ---------------------------------------------------------------------------

export interface SafetySettings {
  no_weapons: boolean;
  no_scary_imagery: boolean;
  no_trademarks: boolean;
  no_stereotypes: boolean;
  age_appropriate_vocab: boolean;
  trademark_safe_prompts: boolean;
  content_sensitivity_precheck: boolean;
  provenance_logging: boolean;
}

export interface CreateChildrensBookRequest {
  title: string;
  subtitle?: string;
  author: string;
  age_range: AgeRange;
  page_count: number;
  trim_size: string;
  illustration_style: IllustrationStyle;
  color_palette: ColorPalette;
  story_mode: StoryMode;
  creation_mode: CreationMode;
  is_bilingual: boolean;
  bilingual_language?: string;
  bilingual_layout?: BilingualLayout;
  fear_intensity: FearIntensity;
  tone: Tone;
  theme_moral?: string;
  main_character?: string;
  setting?: string;
  story_prompt?: string;
  safety_settings: SafetySettings;
}

export interface UpdateChildrensBookRequest {
  title?: string;
  subtitle?: string;
  author?: string;
  age_range?: AgeRange;
  page_count?: number;
  trim_size?: string;
  illustration_style?: IllustrationStyle;
  color_palette?: ColorPalette;
  status?: BookStatus;
}

// ---------------------------------------------------------------------------
// Analysis results
// ---------------------------------------------------------------------------

export interface ReadabilityResult {
  age_range: AgeRange;
  max_sentence_length: number;
  max_word_length: number;
  vocabulary_level: string;
  total_words: number;
  flagged_words: { word: string; reason: string }[];
  flagged_sentences: { page: number; sentence: string; reason: string }[];
  read_aloud_rhythm_score: number;
  page_turn_surprise_map: { page: number; score: number }[];
  look_inside_score: number;
}

export interface RhymeAssistantResult {
  pattern: "AABB" | "ABAB" | "none";
  near_rhymes: { page: number; line: string; suggestion: string }[];
  meter_issues: { page: number; line: string; issue: string }[];
}

export interface ContinuityIssue {
  page_number: number;
  character_name: string;
  issue: string;
  fix_suggestion: string;
}

export interface ContinuityCheckResult {
  score: number;
  issues: ContinuityIssue[];
  style_drift_pages: number[];
}

export interface SafetyCheckResult {
  trademark_flags: { page: number; term: string }[];
  content_flags: { page: number; issue: string; severity: string }[];
  all_clear: boolean;
}

// ---------------------------------------------------------------------------
// Stats
// ---------------------------------------------------------------------------

export interface ChildrensBookStats {
  total_books: number;
  in_progress: number;
  published: number;
  pages_created: number;
}

File: backend/app/modules/specialty/models/childrens.py
"""SQLAlchemy models for Children's Books."""

from __future__ import annotations

import uuid
from typing import Any

from sqlalchemy import (
    JSON,
    Boolean,
    Enum,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    Uuid,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import BaseModel, TenantModel
from app.modules.specialty.models.enums import (
    AgeRange,
    BilingualLayout,
    BookStatus,
    ColorPalette,
    FearIntensity,
    IllustrationStyle,
    PageLayout,
    StoryMode,
    TextPosition,
)


class ChildrensBook(TenantModel):
    """Master record for a children's book."""

    __tablename__ = "childrens_books"

    title: Mapped[str] = mapped_column(String(300), nullable=False)
    subtitle: Mapped[str | None] = mapped_column(String(300), nullable=True)
    author: Mapped[str | None] = mapped_column(String(200), nullable=True)
    age_range: Mapped[str] = mapped_column(
        Enum(AgeRange, name="age_range", native_enum=True),
        nullable=False,
        default=AgeRange.preschool,
    )
    page_count: Mapped[int] = mapped_column(Integer, nullable=False, default=24)
    trim_size: Mapped[str] = mapped_column(String(20), nullable=False, default="8.5x8.5")
    illustration_style: Mapped[str] = mapped_column(
        Enum(IllustrationStyle, name="illustration_style", native_enum=True),
        nullable=False,
        default=IllustrationStyle.watercolor,
    )
    color_palette: Mapped[str] = mapped_column(
        Enum(ColorPalette, name="color_palette", native_enum=True),
        nullable=False,
        default=ColorPalette.bright,
    )
    story_mode: Mapped[str] = mapped_column(
        Enum(StoryMode, name="story_mode", native_enum=True),
        nullable=False,
        default=StoryMode.ai_generated,
    )
    # Fed to the AI story generator and scanned for trademarks; the create and
    # update payloads have always carried them.
    story_prompt: Mapped[str | None] = mapped_column(Text, nullable=True)
    theme_moral: Mapped[str | None] = mapped_column(String(200), nullable=True)
    tone: Mapped[str | None] = mapped_column(String(50), nullable=True)
    is_bilingual: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, server_default="false")
    bilingual_language: Mapped[str | None] = mapped_column(String(50), nullable=True)
    bilingual_layout: Mapped[str | None] = mapped_column(
        Enum(BilingualLayout, name="bilingual_layout", native_enum=True),
        nullable=True,
    )
    fear_intensity: Mapped[str] = mapped_column(
        Enum(FearIntensity, name="fear_intensity", native_enum=True),
        nullable=False,
        default=FearIntensity.none,
    )
    status: Mapped[str] = mapped_column(
        Enum(BookStatus, name="specialty_book_status", native_enum=True),
        nullable=False,
        default=BookStatus.draft,
        server_default="draft",
    )
    qa_score: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Relationships
    pages: Mapped[list[ChildrensBookPage]] = relationship(
        "ChildrensBookPage",
        back_populates="book",
        cascade="all, delete-orphan",
    )
    characters: Mapped[list[ChildrensBookCharacter]] = relationship(
        "ChildrensBookCharacter",
        back_populates="book",
        cascade="all, delete-orphan",
    )


class ChildrensBookPage(BaseModel):
    """Individual page in a children's book."""

    __tablename__ = "childrens_book_pages"

    book_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("childrens_books.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    page_number: Mapped[int] = mapped_column(Integer, nullable=False)
    page_type: Mapped[str] = mapped_column(String(30), nullable=False, default="story")
    layout: Mapped[str] = mapped_column(
        Enum(PageLayout, name="page_layout", native_enum=True),
        nullable=False,
        default=PageLayout.image_top_text_bottom,
    )
    text_content: Mapped[str | None] = mapped_column(Text, nullable=True)
    translated_text: Mapped[str | None] = mapped_column(Text, nullable=True)
    text_font: Mapped[str | None] = mapped_column(String(100), nullable=True)
    text_size: Mapped[int | None] = mapped_column(Integer, nullable=True)
    text_color: Mapped[str | None] = mapped_column(String(20), nullable=True)
    text_position: Mapped[str | None] = mapped_column(
        Enum(TextPosition, name="text_position", native_enum=True),
        nullable=True,
    )
    text_plate_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, server_default="false")
    illustration_prompt: Mapped[str | None] = mapped_column(Text, nullable=True)
    illustration_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    illustration_model: Mapped[str | None] = mapped_column(String(50), nullable=True)
    illustration_seed: Mapped[str | None] = mapped_column(String(50), nullable=True)
    contrast_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    gutter_safe: Mapped[bool | None] = mapped_column(Boolean, nullable=True)
    # Illustration provenance (model, prompt hash, generation date) and upload
    # details, written by generate_illustration and upload_page_image.
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column(JSON, nullable=True)

    # Relationships
    book: Mapped[ChildrensBook] = relationship("ChildrensBook", back_populates="pages")


class ChildrensBookCharacter(BaseModel):
    """Character consistency sheet for a children's book."""

    __tablename__ = "childrens_book_characters"

    book_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("childrens_books.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    species: Mapped[str | None] = mapped_column(String(50), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    reference_images: Mapped[list[str] | None] = mapped_column(JSONB, nullable=True)
    auto_append: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")
    clothing_rules: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)
    scale_rules: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)
    setting_rules: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)
    time_rules: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)

    # Relationships
    book: Mapped[ChildrensBook] = relationship("ChildrensBook", back_populates="characters")
